test/ha/activator_test.go

```diff
@@ -86,73 +92,46 @@ func testActivatorHA(t *testing.T, gracePeriod *int64, slo float64) {
 	)
 	test.EnsureTearDown(t, clients, namesScaleToZero)
 
+	t.Logf("Waiting for %s to scale to zero", namesScaleToZero.Revision)
 	if err := e2e.WaitForScaleToZero(t, revisionresourcenames.Deployment(resourcesScaleToZero.Revision), clients); err != nil {
 		t.Fatal("Failed to scale to zero:", err)
 	}
 
+	t.Log("Starting prober")
 	prober := test.NewProberManager(log.Printf, clients, minProbes)
 	prober.Spawn(resources.Service.Status.URL.URL())
 	defer assertSLO(t, prober, slo)
 
-	pods, err := clients.KubeClient.Kube.CoreV1().Pods(system.Namespace()).List(metav1.ListOptions{
-		LabelSelector: activatorLabel,
-	})
-	if err != nil {
-		t.Fatal("Failed to get activator pods:", err)
-	}
-	activatorPod := pods.Items[0].Name
-
-	origEndpoints, err := pkgTest.GetEndpointAddresses(clients.KubeClient, resourcesScaleToZero.Revision.Name, test.ServingNamespace)
-	if err != nil {
-		t.Fatalf("Unable to get public endpoints for revision %s: %v", resourcesScaleToZero.Revision.Name, err)
-	}
-
-	if err := clients.KubeClient.Kube.CoreV1().Pods(system.Namespace()).Delete(activatorPod, podDeleteOptions); err != nil {
-		t.Fatalf("Failed to delete pod %s: %v", activatorPod, err)
-	}
-
-	// Wait for the killed activator to disappear from the knative service's endpoints.
-	if err := pkgTest.WaitForChangedEndpoints(clients.KubeClient, resourcesScaleToZero.Revision.Name, test.ServingNamespace, origEndpoints); err != nil {
-		t.Fatal("Failed to wait for the service to update its endpoints:", err)
-	}
-
-	assertServiceEventuallyWorks(t, clients, namesScaleToZero, resourcesScaleToZero.Service.Status.URL.URL(), test.PizzaPlanetText1)
-
-	if err := pkgTest.WaitForPodDeleted(clients.KubeClient, activatorPod, system.Namespace()); err != nil {
-		t.Fatalf("Did not observe %s to actually be deleted: %v", activatorPod, err)
+	for i, activator := range activators.Items {
+		t.Logf("Deleting activator%d (%s)", i, activator.Name)
+		if err := clients.KubeClient.Kube.CoreV1().Pods(system.Namespace()).Delete(activator.Name, podDeleteOptions); err != nil {
+			t.Fatalf("Failed to delete pod %s: %v", activator.Name, err)
+		}
+
+		// Wait for the killed activator to disappear from the knative service's endpoints.
+		if err := waitForEndpointsState(clients.KubeClient, resourcesScaleToZero.Revision.Name, test.ServingNamespace, endpointsDoNotContain(activator.Status.PodIP)); err != nil {
+			t.Fatal("Failed to wait for the service to update its endpoints:", err)
+		}
+		if gracePeriod != nil && *gracePeriod == 0 {
+			t.Log("Allow the network to notice the missing endpoint")
+			time.Sleep(5 * time.Second)
```

Comment:
should this wait grace period seconds?

Reply:
You mean the networking grace thingy? 30s? Seemed kinda long.

Reply:
:shrug: